eloop.c

```diff
@@ -484,43 +449,39 @@ int eloop_register_signal_reconfig(eloop_signal_handler handler,
 	return eloop_register_signal(SIGHUP, handler, user_data);
 }
 
+static inline int os_time_to_ms(struct os_time *tv)
+{
+	return ((tv)->sec * 1000 + (tv)->usec / 1000);
+}
 
 void eloop_run(void)
 {
-	fd_set *rfds, *wfds, *efds;
-	int res;
-	struct timeval _tv;
+	int res, timeout = 0;
 	struct os_time tv, now;
-
-	rfds = malloc(sizeof(*rfds));
-	wfds = malloc(sizeof(*wfds));
-	efds = malloc(sizeof(*efds));
-	if (rfds == NULL || wfds == NULL || efds == NULL) {
-		printf("eloop_run - malloc failed\n");
-		goto out;
-	}
+	struct pollfd *fds = NULL;
 
 	while (!eloop.terminate &&
-	       (eloop.timeout || eloop.readers.count > 0 ||
-		eloop.writers.count > 0 || eloop.exceptions.count > 0)) {
+	       (eloop.timeout || eloop.sock_table.count > 0)) {
 		if (eloop.timeout) {
 			os_get_time(&now);
 			if (os_time_before(&now, &eloop.timeout->time))
 				os_time_sub(&eloop.timeout->time, &now, &tv);
 			else
 				tv.sec = tv.usec = 0;
-			_tv.tv_sec = tv.sec;
-			_tv.tv_usec = tv.usec;
+			timeout = os_time_to_ms(&tv);
+		}
+
+		fds = (struct pollfd *)
+			realloc(fds, eloop.sock_table.count * sizeof(struct eloop_sock));
```

Comment:
I believe we are slightly over-allocating memory here. We only need `count * sizeof(struct pollfd)`, not the full `struct eloop_sock` for the `realloc()`.

Reply:
We can adjust this on the merge if @huyizhen agrees.